Exception in Ydoc during shutdown

```
app/gui dev:vite: [ERROR] [2025-10-30T10:53:35+01:00] [org.enso.ydoc.polyfill.web.EventEmitter] Error emitting event of close [[1000, com.oracle.truffle.api.strings.TruffleString]] on () => this.close()
app/gui dev:vite: org.graalvm.polyglot.PolyglotException: java.net.SocketException: Socket closed
app/gui dev:vite: 	at io.helidon.common.buffers@4.2.2/io.helidon.common.buffers.GrowingBufferData.writeTo(GrowingBufferData.java:69)
app/gui dev:vite: 	at io.helidon.common.socket@4.2.2/io.helidon.common.socket.PlainSocket.write(PlainSocket.java:136)
app/gui dev:vite: 	at io.helidon.common.socket@4.2.2/io.helidon.common.socket.SocketWriter.writeNow(SocketWriter.java:81)
app/gui dev:vite: 	at io.helidon.webserver.websocket@4.2.2/io.helidon.webserver.websocket.WsConnection.send(WsConnection.java:342)
app/gui dev:vite: 	at io.helidon.webserver.websocket@4.2.2/io.helidon.webserver.websocket.WsConnection.close(WsConnection.java:205)
app/gui dev:vite: 	at io.helidon.webserver.websocket@4.2.2/io.helidon.webserver.websocket.WsConnection.terminate(WsConnection.java:210)
app/gui dev:vite: 	at org.enso.ydoc.polyfill/org.enso.ydoc.polyfill.web.WebSocket.execute(WebSocket.java:199)
app/gui dev:vite: 	at <js>.close(websocket.js:234)
app/gui dev:vite: 	at <js>.close(ydoc.cjs:28769)
app/gui dev:vite: 	at <js>.:=>(ydoc.cjs:28688)
app/gui dev:vite: 	at org.graalvm.polyglot/org.graalvm.polyglot.Value.executeVoid(Value.java:1074)
app/gui dev:vite: 	at org.enso.ydoc.polyfill/org.enso.ydoc.polyfill.web.EventEmitter$Store.lambda$emit$0(EventEmitter.java:123)
app/gui dev:vite: 	at java.base/java.util.concurrent.ConcurrentHashMap$KeySetView.forEach(ConcurrentHashMap.java:4725)
app/gui dev:vite: 	at org.enso.ydoc.polyfill/org.enso.ydoc.polyfill.web.EventEmitter$Store.emit(EventEmitter.java:120)
app/gui dev:vite: 	at org.enso.ydoc.polyfill/org.enso.ydoc.polyfill.web.EventEmitter.execute(EventEmitter.java:74)
app/gui dev:vite: 	at <js>.emit(event-emitter.js:34)
app/gui dev:vite: 	at <js>.#handle_close(websocket.js:264)
app/gui dev:vite: 	at org.graalvm.polyglot/org.graalvm.polyglot.Value.executeVoid(Value.java:1074)
app/gui dev:vite: 	at org.enso.ydoc.polyfill/org.enso.ydoc.polyfill.web.WebSocket$WebSocketConnection.lambda$onClose$0(WebSocket.java:328)
app/gui dev:vite: 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:545)
app/gui dev:vite: 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:328)
app/gui dev:vite: 	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:309)
app/gui dev:vite: 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1095)
app/gui dev:vite: 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:619)
app/gui dev:vite: 	at java.base/java.lang.Thread.run(Thread.java:1447)
```

--- a/build.sbt
+++ b/build.sbt
@@ -1998,7 +1998,7 @@ lazy val `ydoc-polyfill` = project
     Compile / internalModuleDependencies := Seq(
       (`syntax-rust-definition` / Compile / exportedModule).value
     ),
-    libraryDependencies ++= slf4jApi ++ Seq(
+    libraryDependencies ++= logbackTest ++ slf4jApi ++ Seq(
       "org.graalvm.truffle"  % "truffle-api"                 % graalMavenPackagesVersion % "provided",
       "org.graalvm.polyglot" % "inspect-community"           % graalMavenPackagesVersion % "runtime",
       "org.graalvm.polyglot" % "js-community"                % graalMavenPackagesVersion % "runtime",

--- a/lib/java/ydoc-polyfill/src/main/java/org/enso/ydoc/polyfill/web/WebSocket.java
+++ b/lib/java/ydoc-polyfill/src/main/java/org/enso/ydoc/polyfill/web/WebSocket.java
@@ -9,6 +9,7 @@ import io.helidon.webclient.websocket.WsClientProtocolConfig;
 import io.helidon.webserver.WebServer;
 import io.helidon.webserver.http.HttpRouting;
 import io.helidon.webserver.websocket.WsRouting;
+import io.helidon.websocket.WsCloseCodes;
 import io.helidon.websocket.WsListener;
 import io.helidon.websocket.WsSession;
 import java.io.UncheckedIOException;
@@ -46,6 +47,7 @@ final class WebSocket implements ProxyExecutable {
   private static final String NEW_WEB_SOCKET_SERVER = "new-web-socket-server";
   private static final String WEB_SOCKET_SERVER_START = "web-socket-server-start";
 
+  private static final String SOCKET_CLOSED_REASON = "Socket closed";
   private static final String WEBSOCKET_JS = "websocket.js";
 
   private final ScheduledExecutorService executor;
@@ -196,7 +198,11 @@ final class WebSocket implements ProxyExecutable {
 
         var session = connection.getSession();
         if (session != null) {
-          session.terminate();
+          try {
+            session.terminate();
+          } catch (IllegalStateException socketClosed) {
+            connection.onClose(session, WsCloseCodes.CLOSED_ABNORMALLY, SOCKET_CLOSED_REASON);
+          }
         }
 
         yield null;
@@ -210,7 +216,11 @@ final class WebSocket implements ProxyExecutable {
         var session = connection.getSession();
         if (session != null) {
           var reason = reasonArgument == null ? "Close" : reasonArgument;
-          session.close(code, reason);
+          try {
+            session.close(code, reason);
+          } catch (IllegalStateException socketClosed) {
+            connection.onClose(session, WsCloseCodes.CLOSED_ABNORMALLY, SOCKET_CLOSED_REASON);
+          }
         }
 
         yield null;

--- a/lib/java/ydoc-polyfill/src/main/resources/org/enso/ydoc/polyfill/web/websocket.js
+++ b/lib/java/ydoc-polyfill/src/main/resources/org/enso/ydoc/polyfill/web/websocket.js
@@ -229,6 +229,8 @@
         }
 
         close(code, reason) {
+            if (this.#state == WebSocket.CLOSED) return
+
             this.#state = WebSocket.CLOSING;
             if (code === undefined) {
                 jvm('web-socket-terminate', this._connection);

--- a/lib/java/ydoc-polyfill/src/test/java/org/enso/ydoc/polyfill/web/WebSocketTest.java
+++ b/lib/java/ydoc-polyfill/src/test/java/org/enso/ydoc/polyfill/web/WebSocketTest.java
@@ -249,6 +249,7 @@ public class WebSocketTest extends ExecutorSetup {
           ws.close();
           ws.close();
           res.set(true);
+          // Notify that the client socket is closed
           lock.release();
         });
         """;
@@ -258,6 +259,7 @@ public class WebSocketTest extends ExecutorSetup {
 
     CompletableFuture.supplyAsync(() -> context.eval("js", code), executor).get();
 
+    // Wait until the client socket is closed
     lock.acquire();
 
     Assert.assertTrue(res.get());
@@ -274,11 +276,14 @@ public class WebSocketTest extends ExecutorSetup {
         """
         var ws = new WebSocket('ws://localhost:22334');
         ws.addEventListener('open', () => {
+          // Notify that the client socket is connected
           lock1.release();
+          // Wait until the server socket is closed
           lock2.acquire();
-          // Close WebSocket closed by the server
+          // Close the client socket (closed by the server)
           ws.close();
           res.set(true);
+          // Notify that the client socket is closed
           lock3.release();
         });
         """;
@@ -290,9 +295,13 @@ public class WebSocketTest extends ExecutorSetup {
 
     CompletableFuture.supplyAsync(() -> context.eval("js", code), executor).get();
 
+    // Wait for the client socket connection
     lock1.acquire();
+    // Close the server socket
     ws.stop();
+    // Notify that the server socked is closed
     lock2.release();
+    // Wait until the client socket is closed
     lock3.acquire();
 
     Assert.assertTrue(res.get());

--- a/lib/java/ydoc-polyfill/src/test/java/org/enso/ydoc/polyfill/web/WebSocketTest.java
+++ b/lib/java/ydoc-polyfill/src/test/java/org/enso/ydoc/polyfill/web/WebSocketTest.java
@@ -48,6 +48,7 @@ public class WebSocketTest extends ExecutorSetup {
             .allowAccess(AtomicReference.class.getDeclaredMethod("set", Object.class))
             .allowAccess(
                 AtomicReferenceArray.class.getDeclaredMethod("set", int.class, Object.class))
+            .allowAccess(Semaphore.class.getDeclaredMethod("acquire"))
             .allowAccess(Semaphore.class.getDeclaredMethod("release"))
             .build();
     var contextBuilder = WebEnvironment.createContext(hostAccess);
@@ -232,6 +233,67 @@ public class WebSocketTest extends ExecutorSetup {
     CompletableFuture.supplyAsync(() -> context.eval("js", code), executor).get();
 
     lock.acquire();
+
+    Assert.assertTrue(res.get());
+  }
+
+  @Test
+  public void closeIdempotent() throws Exception {
+    var lock = new Semaphore(0);
+    var res = new AtomicBoolean(false);
+
+    var code =
+        """
+        var ws = new WebSocket('ws://localhost:22334');
+        ws.addEventListener('open', () => {
+          ws.close();
+          ws.close();
+          res.set(true);
+          lock.release();
+        });
+        """;
+
+    context.getBindings("js").putMember("lock", lock);
+    context.getBindings("js").putMember("res", res);
+
+    CompletableFuture.supplyAsync(() -> context.eval("js", code), executor).get();
+
+    lock.acquire();
+
+    Assert.assertTrue(res.get());
+  }
+
+  @Test
+  public void handleSocketClosed() throws Exception {
+    var lock1 = new Semaphore(0);
+    var lock2 = new Semaphore(0);
+    var lock3 = new Semaphore(0);
+    var res = new AtomicBoolean(false);
+
+    var code =
+        """
+        var ws = new WebSocket('ws://localhost:22334');
+        ws.addEventListener('open', () => {
+          lock1.release();
+          lock2.acquire();
+          // Close WebSocket closed by the server
+          ws.close();
+          res.set(true);
+          lock3.release();
+        });
+        """;
+
+    context.getBindings("js").putMember("lock1", lock1);
+    context.getBindings("js").putMember("lock2", lock2);
+    context.getBindings("js").putMember("lock3", lock3);
+    context.getBindings("js").putMember("res", res);
+
+    CompletableFuture.supplyAsync(() -> context.eval("js", code), executor).get();
+
+    lock1.acquire();
+    ws.stop();
+    lock2.release();
+    lock3.acquire();
 
     Assert.assertTrue(res.get());
   }

